File: kyulee/programmers-2/스킬트리.java
import java.util.*;

public class 스킬트리 {
    class Solution {
        public int solution(String skill, String[] skill_trees) {
            int answer = 0;

            for (String tree : skill_trees) {
                if (isPossibleSkillTree(skill, tree)) {
                    answer++;
                }
            }

            return answer;
        }

        private boolean isPossibleSkillTree(String skill, String tree) {
            // 주어진 스킬트리에서 선행 스킬 순서에 맞게 스킬을 배웠는지 확인
            int idx = 0;
            for (char c : tree.toCharArray()) {
                if (skill.indexOf(c) != -1) {
                    if (skill.charAt(idx) == c) {
                        idx++;
                    } else {
                        return false;
                    }
                }
            }
            return true;
        }
    }
}

File: soyeong/programmers-2/스킬트리.py
# skill = 선행 스킬
def solution(skill, skill_trees):
    answer = 0
    for tree in skill_trees:
        s = ''
        for i in tree:
            # 스킬트리안 스킬을 선행스킬과 관련된 것만으로 정리하는 과정
            # (여기서 스킬 트리와 상관 없이 배울 수 있는 스킬은 검증 대상에서 제외되고 해당 스킬들만 정리)
            if i in skill:
                s += i

        # 선행 스킬에서 정리된 스킬트리와 대조했을때 같다면 습득 가능한 스킬트리 이므로 +1
        if skill[:len(s)] == s:
            answer += 1

    return answer